### DISCUSSION TITLE
Procedural skeletal animation

### DISCUSSION THREAD

**Discussion by LeXXik** [jrouwe/JoltPhysics.js#122] [Q&A] [answered]
I am trying to create a `SkeletalAnimation` for a procedural animation. By procedural I mean that the final keyframes of bones transforms are not available offline, and are generated procedurally. For example, we have animation layers, where each layer is playing a different animation (each is imported from GLB/FBX and of different duration) and the engine then blends them to create a final seamless animation (e.g. 2 layers: lower body is running (legs), upper body is boxing).

Any suggestions on how to approach it? Should I create a skeletal animation for a single keyframe (bone transforms for the current frame only) and then update those keyframes each frame before driving the pose to reach them?

**Comment by LeXXik** [comment]
Tried to create 2 keyframes and on each update:

1. write pos/rot from the second to the first one
2. write new pos/rot to the second
3. set first time to zero, second to current dt
4. `pose.CalculateJointMatrices`
5. `ragdoll.DriveToPoseUsingKinematics`

Gets launched into space....

**Comment by jrouwe** [comment]
I would not use a SkeletalAnimation at all, I would use the SkeletonPose directly. You can either use GetJoint and update the local space transform or you can use GetJointMatrix and update the model space transform directly.

**Comment by LeXXik** [comment]
Hmm, so here is my attempt so far.

Firstly, the launch to space was due to my setup, where the rendering is decoupled from the physics update. E.g. there would be about 2 browser frames in a single physics step. So, calling `ragdoll.DriveToPoseUsingKinematics(pose, dt)` on browser update with `dt` being frame's dt would cause it to be called twice per simulation step (albeit with lower dt). Syncing it with the physics update and setting the `dt` to that of the physics update, prevented the ragdoll from flying off.

Here, in the video clip, I am using 2 different animations blended together to get the final one. As per recommendation, I am now not creating an animation skeleton, but using the pose joints directly by looping them and setting the local transforms on them. I am having trouble with a powered ragdoll.

In the video clip, I show 3 setups in their order (mesh view is the physics debug draw, blue kinematics, yellow dynamics). On each update I update the joints transforms and call `pose.CalculateJointMatrices()`:

1. All skeleton parts are kinematic. After transforms update I call `ragdoll.DriveToPoseUsingKinematics(pose, fixedStep)`.
![image](https://github.com/jrouwe/JoltPhysics.js/assets/5677782/2aa1e212-485a-494f-8743-de64046faad3)

2. All parts are dynamic. Same update, but call `ragdoll.SetPose(pose)` instead.
![image](https://github.com/jrouwe/JoltPhysics.js/assets/5677782/9356bde5-b836-42f6-bbfa-301e98493dcc)

3. All parts are dynamic. Same update, but call `ragdoll.DriveToPoseUsingMotors(pose)`.
![image](https://github.com/jrouwe/JoltPhysics.js/assets/5677782/1d507868-19ad-4944-8c9b-8b6b44302db7)

I suppose when I call `SetPose(pose)`, it just teleports the parts to their respected positions, disregarding the environment. When asking to drive them via motors, though, it seems like it does try to do so (parts are moving), but not enough juice? Not sure. I tried to increase the limits of motors, but that didn't seem to give any effect.

**Reply by jrouwe** [reply, marked as answer]
So first of all, for case 3 I would expect the character to fall over immediately because the motors work in local space and there is no motor on the root joint to hold it up. If you don't want that then make the first body kinematic and the rest dynamic or use an extra constraint to constrain the root to the world. Other than that the joints indeed seem to be not powerful enough. I assume you're using a SwingTwistConstraint? Use the default settings for mSwingMotorSettings and mTwistMotorSettings but try setting `mSpringSettings.mFrequency = 20` for both.

**Reply by LeXXik** [reply]
Thank you! Your comment gave me confidence that I am not doing something completely wrong. Part of an issue, I realized, was that I had too few sub steps for the constraints to be rigid enough. I think I can use position and velocity step overrides on the constraints themselves, so that I don't increase global sub steps. The other issue is that I think some of the normals or the angles of the swing-twist constraint are not correct or too limited. 

Just to make sure I understand the swing-twist image from the docs correctly, if we apply it on a person:

- plane axis: pointing outwards from the chest
- twist axis: left arm outstretched to the side
- normal cone angle / 2: arm flapping up/down like a bird wing
- plane cone angle / 2: arm moving forward/back angle
- twist angle: arm staying in place, twisting like a skrewdriver
- normal axis: auto calculated, ortho to plane axis and twist axis

If I want to know whether the animation is straining the angle of some constraint (e.g. constantly trying to push outside its min/max limits), what would be the best way to find it? I can get the constraint from ragdoll, but it doesn't seem like I can read the current e.g. normal cone half angle.

**Reply by jrouwe** [reply]
> Part of an issue, I realized, was that I had too few sub steps for the constraints to be rigid enough

The default of 10 velocity and 2 position steps should be plenty for a strong constraint.

> Just to make sure I understand the swing-twist image from the docs correctly

That sounds correct.

> I can get the constraint from ragdoll, but it doesn't seem like I can read the current e.g. normal cone half angle.

There is `SwingTwistConstraint::GetRotationInConstraintSpace()` which gives you a quaternion which you can decompose using `Quat::GetSwingTwist()` followed by `Quat::GetRotationAngle()` where X-axis is twist and YZ swing. But a simpler way may be to call `GetTargetOrientationCS` after you did `SetTargetOrientationCS` and if the quaternion differs significantly then it was clamped, or you could just open up the constraint completely by setting twist min = -pi, twist max = pi, half cone = pi, half plane = pi and see where it ends up.

**Comment by LeXXik** [comment]
> The default of 10 velocity and 2 position steps should be plenty for a strong constraint.
Right. I had it both on 2 by default, so the velocity was a bit low.

Thank you for the advice on debugging the angles. My latest attempt:

1/30 dt
2 substeps
10 velocity
2 position
10 damping
30 frequency

https://github.com/jrouwe/JoltPhysics.js/assets/5677782/848cd018-1c06-4bd8-a061-db1aa7ff1d10

A bit springy, so will play with params some more, maybe increase damping to tighten it. I will mark this as answered.
